hack/go-version-check.sh

```diff
@@ -0,0 +1,113 @@
+#!/bin/bash
+#
+# List or check that the golang version embedded in files (specifically our
+# Dockerfile) is using a supported version of Go.
+# Run with --show to list the images the script thinks are current.
+# Run with --check (the default) to verify the Dockerfile contains a current
+# image & version tag.
+#
+
+set -e
+SCRIPT_DIR="$(readlink -f "$(dirname "${0}")")"
+PROJECT_DIR="$(readlink -f "${SCRIPT_DIR}/..")"
+WORKDIR="$(mktemp -d)"
+CONTAINER_FILES=("${PROJECT_DIR}/Dockerfile")
+GOVERSURL='https://go.dev/dl/?mode=json'
+ACTION=check
+
+
+fetch_versions() {
+    local url="$1"
+    local out="$2"
+    if [ -f "${out}" ]; then
+        return
+    fi
+    echo "Fetching Go versions..." >&2
+    curl --fail -sL -o "${out}" "${url}"
+}
+
+extract_versions() {
+    local jsonfile="$1"
+    jq -r '.[].version' < "${jsonfile}" | \
+        cut -d. -f1,2 | sort -u | sed -e 's,^go,,'
+}
+
+to_images() {
+    for a in "$@"; do
+        echo "docker.io/golang:${a}"
+    done
+}
+
+
+opts=$(getopt --name "$0" \
+    -o "csf:" -l "check,show,file:,url:,versionsfile:" -- "$@")
+eval set -- "$opts"
+
+cli_cfiles=()
+while true ; do
+    case "$1" in
+        -c|--check)
+            ACTION=check
+            shift
+        ;;
+        -s|--show)
+            ACTION=show
+            shift
+        ;;
+        -f|--file)
+            cli_cfiles+=("$2")
+            shift 2
+        ;;
+        --url)
+            GOVERSURL="$2"
+            shift 2
+        ;;
+        --versionsfile)
+            GOVERSIONSFILE="$2"
+            shift 2
+        ;;
+        --)
+            shift
+            break
+        ;;
+        *)
+            echo "unexpected option: $1" >&2
+            exit 2
+        ;;
+    esac
+done
+
+if [ "${#cli_cfiles}" -gt 0 ]; then
+    CONTAINER_FILES=("${cli_cfiles[@]}")
+fi
+
+cleanup() {
+    # only used in trap, ignoe unreachable error
+    # shellcheck disable=SC2317
+    rm -rf "${WORKDIR}"
+}
+trap cleanup EXIT
```

Comment:
To catch the exit for any unexpected option we need to place the `trap` well before those command-line parsing statements from above. Otherwise we end up with empty temporary directories.

Reply:
Right, done.